lib/charms/opensearch/v0/opensearch_relation_provider.py

```diff
@@ -0,0 +1,254 @@
+
+        if permissions or action_groups:
+            # combine action groups and perms into a new role for this relation.
+            try:
+                self.user_manager.create_role(
+                    role_name=username,
+                    permissions=permissions,
+                    action_groups=action_groups,
+                )
+                roles.append(username)
+            except OpenSearchUserMgmtError as err:
+                logger.error(err)
+                raise
+
+        try:
+            self.user_manager.create_user(
+                username,
+                roles,
+                hashed_pwd,
+            )
+            if roles:
+                self.user_manager.patch_user(
+                    username,
+                    [{"op": "replace", "path": "/opendistro_security_roles", "value": roles}],
+                )
+        except OpenSearchUserMgmtError as err:
+            logger.error(err)
+            raise
+
+    def _on_relation_departed(self, event: RelationDepartedEvent) -> None:
+        """Check if this relation is being removed, and update the peer databag accordingly."""
+        if event.departing_unit == self.charm.unit:
+            self.charm.peers_data.put(Scope.UNIT, self._depart_flag(event.relation), True)
+
+    def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Handle client relation-broken event."""
+        if not self.unit.is_leader():
+            return
+        if self._unit_departing(event.relation):
+            # This unit is being removed, so don't update the relation.
+            self.charm.peers_data.delete(Scope.UNIT, self._depart_flag(event.relation))
+            return
+
+        username = self._relation_username(event.relation)
+        try:
+            self.user_manager.remove_user(username)
+        except OpenSearchUserMgmtError as err:
+            logger.error(f"failed to remove role {username}: {str(err)}")
+            lingering_users = self.charm.peers_data.get(Scope.APP, "lingering_users").split(",")
+            lingering_users.add(username)
+            self.charm.peers_data.put(Scope.APP, "lingering_users", ",".join(lingering_users))
+
+        try:
+            self.user_manager.remove_role(username)
+        except OpenSearchUserMgmtError as err:
+            logger.error(f"failed to remove role {username}: {str(err)}")
+            lingering_roles = self.charm.peers_data.get(Scope.APP, "lingering_roles").split(",")
+            lingering_roles.append(username)
+            self.charm.peers_data.put(Scope.APP, "lingering_roles", (lingering_roles))
+
+    def clear_lingering_users_and_roles(self) -> None:
+        """Remove users and roles that we failed to delete when removing a relation."""
+        if not self.opensearch.is_node_up() or not self.unit.is_leader():
+            return
+
+        lingering_users = set(
+            self.charm.peers_data.get(Scope.APP, "lingering_users", "").split(",")
+        )
```

Comment:
it is unclear why peer databag magic is needed 😞 
better implementation https://github.com/canonical/mongodb-operator/blob/main/lib/charms/mongodb/v0/mongodb_provider.py#L124-L126